### PR TITLE
Ho model

### PR DESCRIPTION
**Pull request by ajreynol**
This extends CVC4's model construction algorithm for higher-order.

The fundamental change is that we must consider a "dag form" model representation for uninterpreted functions instead of a "table form", or else function representations can be exponential in size.  The following example demonstrates this issue:

(set-logic UFLIA)
(set-info :status sat)
(declare-fun f1 (Int Int Int Int) Int)
(declare-fun f2 (Int Int Int) Int)
(declare-fun f3 (Int Int) Int)
(declare-fun f4 (Int) Int)

(assert (= (f1 0) (f1 1)))
(assert (= (f1 1) f2))

(assert (= (f2 0) (f2 1)))
(assert (= (f2 1) f3))

(assert (= (f3 0) (f3 1)))
(assert (= (f3 1) f4))

(assert (= (f4 0) (f4 1)))
(assert (= (f4 1) 2))

...

Notice that all applications apart from f4 are partial.  This is equivalent to saying that "all combinations of 0 and 1 as arguments for f1 return the value 2".

With a table construction, we generate the model for f1:

(lambda xyzw.
(ite (and (= x 0) (= y 0) (= z 0) (= w 0)) 2
(ite (and (= x 0) (= y 0) (= z 0) (= w 1)) 2
(ite (and (= x 0) (= y 0) (= z 1) (= w 0)) 2
(ite (and (= x 0) (= y 0) (= z 1) (= w 1)) 2
...
(ite (and (= x 1) (= y 1) (= z 1) (= w 1)) 2
...))))))

which is exponential with respect to the number of input constraints.

With a tree (dag) construction, we generate the model for f1 in terms of a dag:

(lambda xyzw.
(let ((f4v (ite (= w 0) 2 (ite (= w 1) 2 ...))))
(let ((f3v (ite (= z 0) f4v (ite (= z 1) f4v ...))))
(let ((f2v (ite (= y 0) f3v (ite (= y 1) f3v ...))))
(ite (= x 0) f2v (ite (= x 1) f2v ...))))))

This remains proportional to the number of input constraints.

To summarize the changes to the model in this commit:

- The getValue function is different for functions now (it may consult the equality engine for representative if in higher-order).
- We use a dag model construction if higher-order is enabled.  We maintain a list of HO_APPLY terms in addition to APPLY_UF terms in this case.
- Assign function definition is now encapsulated and also sets d_uf_model for functions in the same equivalence class when higher-order is enabled.
- The debugCheckModel function was modified to be more robust (it previously was incompatible with functions since it used an auxiliary d_constantReps map instead of querying the representative).

### CODE REVIEW

**Comment by ajreynol** [comment]
This commit is 7/10 planned commits from the ho-smt branch: https://github.com/CVC4/CVC4/tree/ho-smt

**Comment by ajreynol** [comment]
Btw, regarding the special case in:
https://github.com/CVC4/CVC4/pull/1120/files#diff-c473e454b74828446a5f7030f7840815R220
An alternative I considered was to extend the equality engine with a public function ee.hasExternalTerm( n ) which returns true iff ee.hasTerm( n ) is true *and* n is not marked as an internal node in equality engine ee.  
This would avoid the special case here but adds an additional function to the interface of the equality engine which might be confusing.

File: src/theory/theory_model.h [hunk 1]
```diff
@@ -130,12 +130,26 @@ class TheoryModel : public Model
...
   void printRepresentative( std::ostream& out, Node r );
+private:
+  /** uf models */
+  std::map< Node, Node > d_uf_models;
 public:
   /** whether function models are enabled */
   bool d_enableFuncModels;
-  //necessary information for function models
+  /** a map from functions f to a list of all APPLY_UF terms with operator f */
   std::map< Node, std::vector< Node > > d_uf_terms;
```
Comment (barrettcw) [inline src/theory/theory_model.h]: Can you use unordered maps here and below?

File: src/theory/theory_model.h [hunk 2]
```diff
@@ -130,12 +130,26 @@ class TheoryModel : public Model
...
-  std::map< Node, Node > d_uf_models;
+  /** a map from functions f to a list of all HO_APPLY terms with first argument f */
+  std::map< Node, std::vector< Node > > d_ho_uf_terms;
+  /** assign function value f to definition f_def */
+  void assignFunctionDefinition( Node f, Node f_def );
+  /** have we assigned function f? */
+  bool hasAssignedFunctionDefinition( Node f ) const { return d_uf_models.find( f )!=d_uf_models.end(); }
+  /** get the list of functions to assign.
```
Comment (barrettcw) [inline src/theory/theory_model.h]: Please be more specific.  What are the criteria for the functions in the list and how is the list obtained?

File: src/theory/theory_model.h [hunk 3]
```diff
@@ -307,6 +321,44 @@ class TheoryEngineModelBuilder : public ModelBuilder
...
+  * (f 0) = (lambda y. (ite (= y 1) 1
+  *                    (ite (= y 2) 2 ...))
+  * (f 1) = (lambda y. (ite (= y 1) 3 ...))
+  * where
+  * f = (lambda xy. (ite (= x 0) ((f 0) y)
+  *                 (ite (= x 1) ((f 1) y) ...))
+  */
+  void assignHoFunction(TheoryModel* m, Node f);
+  /** assign all unassigned functions that occur in the model m, using the two functions above. */
+  void assignFunctions(TheoryModel* m);
```
Comment (barrettcw) [inline src/theory/theory_model.h]: How do you determine which of the above functions to use?

File: src/theory/theory_model.h [hunk 4]
```diff
@@ -130,12 +130,26 @@ class TheoryModel : public Model
   void printRepresentativeDebug( const char* c, Node r );
   /** print representative function */
   void printRepresentative( std::ostream& out, Node r );
+private:
+  /** uf models */
```
Comment (barrettcw) [inline src/theory/theory_model.h]: Explain more; what is this a map from and what is it a map to.

File: src/theory/theory_model.cpp [hunk 5]
```diff
@@ -233,9 +213,46 @@ Node TheoryModel::getModelValue(TNode n, bool hasBoundVars, bool useDontCares) c
...
-
-    if (!d_equalityEngine->hasTerm(n)) {
-      if(n.getType().isRegExp()) {
+  
+    Debug("model-getvalue-debug") << "Handling special cases for types..." << std::endl;
+    TypeNode t = n.getType();
+    bool eeHasTerm;
+    if( !options::ufHo() && (t.isFunction() || t.isPredicate()) ){
+      // functions are in the equality engine, but *not* as first-class members
+      // when higher-order is disabled. In this case, we cannot not query representatives for functions
```
Comment (barrettcw) [inline src/theory/theory_model.cpp]: cannot not?

File: src/theory/theory_model.cpp [hunk 6]
```diff
@@ -1107,44 +1250,161 @@ bool TheoryEngineModelBuilder::preProcessBuildModel(TheoryModel* m) {
...
     }
   }
-  return true;
+  Node val = NodeManager::currentNM()->mkNode( kind::LAMBDA, 
+               NodeManager::currentNM()->mkNode( kind::BOUND_VAR_LIST, args ), curr );
+  m->assignFunctionDefinition( f, val );
+}
+
+
+struct sortTypeSize {
```
Comment (barrettcw) [inline src/theory/theory_model.cpp]: This should have a comment (like a class).

File: src/theory/theory_model.cpp [hunk 7]
```diff
@@ -1107,44 +1250,161 @@ bool TheoryEngineModelBuilder::preProcessBuildModel(TheoryModel* m) {
...
+  std::map< TypeNode, unsigned > d_type_size;
+  unsigned getTypeSize( TypeNode tn ) {
```
Comment (barrettcw) [inline src/theory/theory_model.cpp]: method comment

File: src/theory/theory_model.cpp [hunk 8]
```diff
@@ -1107,44 +1250,161 @@ bool TheoryEngineModelBuilder::preProcessBuildModel(TheoryModel* m) {
...
+      unsigned sum = 1;
+      for( unsigned i=0; i<tn.getNumChildren(); i++ ){
+        sum += getTypeSize( tn[i] );
+      }  
+      d_type_size[tn] = sum;
+      return sum;
+    }
+  }
+public:
+  bool operator() (Node i, Node j) {
```
Comment (barrettcw) [inline src/theory/theory_model.cpp]: method comment

**Comment by ajreynol** [comment]
I've improved the comments in my latest commit

File: src/theory/theory_model.h (hunk 1 above)
Comment (ajreynol) [reply src/theory/theory_model.h]: These should probably stay as std::map for now since I iterate over them for finite model finding.

**Comment by ajreynol** [comment]
The other portions of TheoryModel are not documented currently, I've opened #1199.
